### ISSUE TITLE
Some keypresses triggering twice since 2.28.2

### ISSUE THREAD

**Issue by barbeque-squared** [libsdl-org/SDL#8193]
How to reproduce:
- Use Archlinux (I'm assuming other distro's are also affected, but can't verify that)
- Install [sdl2](https://archlinux.org/packages/extra/x86_64/sdl2/) version 2.28.2 or higher
- Install or build [USDX](https://github.com/UltraStar-Deluxe/USDX)
- Extract this test song in `~/.ultrastar/songs` (if you've installed it) or in `game/songs` (if you've built it): [Drum Beat.zip](https://github.com/libsdl-org/SDL/files/12505804/Drum.Beat.zip)
- Follow the instructions in ticket https://github.com/UltraStar-Deluxe/USDX/issues/743

What I've tested:
- It doesn't matter which version of SDL is installed when building USDX
- Behaviour has been correct for at least the last 4 years
- Incorrect behaviour occurs when SDL 2.28.2 or 2.28.3 is installed
- Rolling back to 2.28.1 and restarting USDX (no recompilation needed!) immediately fixes the issue. _This leads me to believe it's an issue or at least a change in SDL_

Details:
I don't exactly know how USDX handles input, but searching its codebase for `SDLK_KP_MINUS` should give the most relevant places. Other relevant places are most likely found by searching for `Display.ParseInput` and `SDL_KEYDOWN, SDL_TEXTINPUT`

It's possible that USDX is handling input in an outdated manner, however I don't see any immediate relation between the 2.28.2 release notes and this sudden change in behaviour.

It's probably possible to create a much smaller way of reproducing this bug.

**Comment by smcv** [comment]
X11 or native Wayland? (The default in SDL 2 is X11, even if you're in a Wayland environment, so it's X11 unless you have explicitly set `SDL_VIDEODRIVER=wayland`.)

This is genuinely keyboard input and not a game controller, am I correct?

3f93f56b could conceivably have triggered this; does reverting that commit resolve this for you? (That would only be a workaround rather than a solution, but it would give maintainers a data point.)

I don't see anything else in the diff between 2.28.1 and 2.28.2 that would affect keyboard input.

**Comment by barbeque-squared** [comment]
@smcv 
> X11 or native Wayland?

X11

> This is genuinely keyboard input and not a game controller, am I correct?

Actual keyboard input. I don't have any controllers connected. `xev` shows `KeyPress` events from both my regular keyboard as well as my separate numpad, not sure if there's some better command to verify this.

> does reverting that commit (https://github.com/libsdl-org/SDL/commit/3f93f56b6eea097597bd7ea54bb69642af0141ff) resolve this for you?

No, reverting that commit does not fix it. This is using the arch `sdl2-git` package and adding a `(cd SDL && git revert 3f93f56b6eea097597bd7ea54bb69642af0141ff)` as the first step of the `build()` section.

I tried building it using various `git checkout` which builds, but it breaks things because some .so names are then different. It would be fairly easy to revert whole lists of commit sha1's if I knew which ones don't change what .so gets generated, and narrow things down from there. I assume that to start with it would be all commits that only change stuff in some directory (and not anywhere else) but I don't know exactly what criteria I'm looking for here.

**Comment by smcv** [comment]
If you say SDL 2.28.1 is "good" for this bug, and 2.28.2 is "bad", then the complete changelog between those versions is:

https://github.com/libsdl-org/SDL/compare/release-2.28.1...release-2.28.2

You should be able to `git checkout` any commit between those points and find that it identifies itself as 2.28.1 (or close enough), and they should all be fully compatible with each other.

Compiling the release-2.28.1 tag and the release-2.28.2 tag would be a good starting point if you haven't already, to confirm whether the upstream source code, when built with your current compiler, options etc., still has or doesn't have the bug. After that, I'd suggest bisecting, either with `git bisect` or manually (pick a point about half way between the newest known-good commit and the oldest known-bad commit, test it, repeat until you have narrowed down to a good commit that is immediately followed by a bad commit).

But, if you're using `sdl2-git` from Arch AUR, that is not actually any point on the `release-2.28.x` branch: it seems to be tracking the SDL2 branch, which is the branch that will eventually produce SDL 2.29.x prereleases leading to a 2.30.0 release. Were you definitely testing 2.28.1 and 2.28.2 before, and not the SDL2 branch?

> I assume that to start with it would be all commits that only change stuff in some directory (and not anywhere else)

I tried to narrow it down like that, and I couldn't see any commits other than 3f93f56b6eea097597bd7ea54bb69642af0141ff that looked like they might have triggered this. So if reverting that commit didn't help, clearly I was wrong, and the trigger must be something that I didn't think could cause this.

**Comment by barbeque-squared** [comment]
I tried a bunch of different commits/branches, some don't build, but I was able to build some really old versions (I think 2.26.0) that also still exhibited the bugged behaviour -- which makes no sense at all.

Maybe it's something else _in combination_ with the newer SDL's, but I have no idea what.

What I did try was try some different variations of the line in the USDX source code that you find by searching for `SDL_KEYDOWN, SDL_TEXTINPUT`.
If I remove TEXTINPUT: a lot of keys stop working
If I remove KEYDOWN: the arrow keys/Enter stop working (mouse navigation still works) but the editor bug is actually fixed
If I have both: the bug happens.

Just looking at the USDX code makes me wonder how this ever even remotely worked in the first place. It's annoying that I can't identify the specific commit that actually causes the new/bugged behaviour, but it's probably easier to just redo the way USDX handles input at this point. I've had some off-Github reports from _some_ Windows users that describes an extremely similar problem, but the SDL versions aren't lining up.

It's fine if you want to close this issue.
